### PR TITLE
HPCC-28690 Blobs do not work in indexes created with COMPRESSED(FIRST) option

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Using a trivial index with COMPRESSED(FIRST) on it

### CODE REVIEW

File: system/jhtree/ctfile.cpp [hunk 1]
```diff
@@ -324,7 +324,7 @@ bool CWriteNode::add(offset_t pos, const void *indata, size32_t insize, unsigned
         keyPtr += sizeof(rsequence);
         hdr.keyBytes += sizeof(rsequence);
     }
-    if (isLeaf() && (keyType & HTREE_COMPRESSED_KEY))
+    if (isBlob() || (isLeaf() && (keyType & HTREE_COMPRESSED_KEY)))
```
Comment (richardkchapman) [inline system/jhtree/ctfile.cpp:327]: I'm pretty sure this test is spurious as writing of blob nodes does not go via this class. I will check.

File: system/jhtree/ctfile.cpp [hunk 2]
```diff
@@ -284,7 +284,7 @@ void CWriteNodeBase::writeHdr()
 
 void CWriteNodeBase::write(IFileIOStream *out, CRC32 *crc)
 {
-    if (isLeaf() && (keyHdr->getKeyType() & HTREE_COMPRESSED_KEY))
+    if (isBlob() || (isLeaf() && (keyHdr->getKeyType() & HTREE_COMPRESSED_KEY)))
```
Comment (richardkchapman) [inline system/jhtree/ctfile.cpp:287]: I think this is the fix for the corrupted info on writing blobs in a non-compressed index

File: system/jhtree/ctfile.cpp [hunk 3]
```diff
@@ -1111,20 +1111,8 @@ void CJHTreeBlobNode::unpack(const void *nodeData, bool needCopy)
     CJHTreeNode::unpack(nodeData, needCopy);
     const byte *data = ((const byte *) nodeData) + sizeof(hdr);
     char keyType = keyHdr->getKeyType();
-    if (keyType & HTREE_COMPRESSED_KEY)
```
Comment (richardkchapman) [inline system/jhtree/ctfile.cpp]: This fixes the reading side

File: system/jhtree/ctfile.cpp (hunk 1 above)
Comment (richardkchapman) [reply system/jhtree/ctfile.cpp]: Checked - it was spurious and I have repushed without it

**Review by ghalliday — APPROVED** [review]

**Pull request merged**